xnmt/transformer.py

```diff
@@ -0,0 +1,315 @@
+
+  def __call__(self, input_expr, reconstruct_shape=True, timedistributed=False):
+    W1 = dy.parameter(self.W1)
+    b1 = dy.parameter(self.b1)
+
+    if not timedistributed:
+        input = TimeDistributed()(input_expr)
+    else:
+        input = input_expr
+    output = dy.affine_transform([b1, W1, input])
+
+    if not reconstruct_shape:
+        return output
+    (_, seq_len), batch_size = input_expr.dim()
+    return ReverseTimeDistributed()(output, seq_len, batch_size)
+
+
+class Linear_nobias(object):
+  def __init__(self, dy_model, input_dim, output_dim):
+    self.W1 = dy_model.add_parameters((output_dim, input_dim), init=linear_init(input_dim))
+    self.output_dim = output_dim
+
+  def __call__(self, input_expr):
+    W1 = dy.parameter(self.W1)
+    b1 = dy.zeros(self.output_dim)
+    (_, seq_len), batch_size = input_expr.dim()
+
+    output = dy.affine_transform([b1, W1, input_expr])
+
+    if seq_len == 1: # This is helpful when sequence length is 1, especially during decoding
+        output = ReverseTimeDistributed()(output, seq_len, batch_size)
+    return output
+
+
+class LayerNorm(object):
+  def __init__(self, dy_model, d_hid):
+    self.p_g = dy_model.add_parameters(dim=d_hid)
+    self.p_b = dy_model.add_parameters(dim=d_hid)
+
+  def __call__(self, input_expr):
+    # g = dy.parameter(self.p_g)
+    # b = dy.parameter(self.p_b)
+    #
+    # (_, seq_len), batch_size = input_expr.dim()
+    # input = TimeDistributed()(input_expr)
+    # output = dy.layer_norm(input, g, b)
+    # return ReverseTimeDistributed()(output, seq_len, batch_size)
+    return input_expr
+
+
+def split_rows(X, h):
+  (n_rows, _), batch = X.dim()
+  l = range(n_rows)
+  steps = n_rows // h
+  output = []
+  for i in range(0, n_rows, steps):
+    output.append(dy.pickrange(X, i, i + steps))
+  return output
+
+
+def split_batch(X, h):
+  (n_rows, _), batch = X.dim()
+  l = range(batch)
+  steps = batch // h
+  output = []
+  for i in range(0, batch, steps):
+    indexes = l[i:i + steps]
+    output.append(dy.pick_batch_elems(X, indexes))
+  return output
+
+
+class MultiHeadAttention(object):
+  """ Multi Head Attention Layer for Sentence Blocks
+  For batch computation efficiency, dot product to calculate query-key
+  scores is performed all heads together.
+  """
+  def __init__(self, dy_model, n_units, h=1, self_attention=True, attn_dropout=False):
+    self.W_Q = Linear_nobias(dy_model, n_units, n_units)
+    self.W_K = Linear_nobias(dy_model, n_units, n_units)
+    self.W_V = Linear_nobias(dy_model, n_units, n_units)
+    self.finishing_linear_layer = Linear_nobias(dy_model, n_units, n_units)
+    self.h = h
+    self.scale_score = 1. / (n_units // h) ** 0.5
+    self.is_self_attention = self_attention
+    self.attn_dropout = attn_dropout
+
+  def set_dropout(self, dropout):
+    self.dropout = dropout
+
+  def __call__(self, x, z=None, mask=None):
+    h = self.h
+    if self.is_self_attention:
+      Q = self.W_Q(x)
```

Comment:
I took another look at this, and I still think that this is incorrect. Look at the equation at the end of Section 3.2.2: https://arxiv.org/pdf/1706.03762.pdf

It's not doing a single matrix multiply over the entire input, but instead `h` individual matrix multiplies over each of the segments of the input.

To connect what you're doing with what they're doing in the paper: your matrix is a regular matrix with all the parameters being non-zero. Their matrix is a block diagonal matrix (https://en.wikipedia.org/wiki/Block_matrix#Block_diagonal_matrices), where all of the parameters that are off the diagonal are zero.

So basically, I think you should implement attention similar to the formulas in their paper just using for loops for now. We could consider a more efficient way to do this with tensor contractions later.